blockonomics.php

```diff
@@ -309,71 +353,78 @@ public function testSetup()
-            $response = $this->getNewAddress(true);
-            if ($response->response_code != 200) {
-                $error_str = $response->data->message;
-            }
+        return $error_str;
+    }
+
+    public function checkGetCallbacksResponseBody($response, $crypto)
+    {
+        $error_str = '';
+
+        if (!isset($response->data) || count($response->data) == 0) {
+            $error_str = $this->l("Please add a new store on blockonomics' website");
+        } elseif (count($response->data) >= 1) {
+            $error_str = $this->examineServerCallbackUrls($response->data, $crypto);
         }
+        return $error_str;
+    }
 
+    // checks each existing xpub callback URL to update and/or use
+    public function examineServerCallbackUrls($response_body, $crypto)
+    {
+        $callback_secret = Configuration::get('BLOCKONOMICS_CALLBACK_SECRET');
+        $api_url = Context::getContext()->shop->getBaseURL(true) . 'modules/' . $this->name;
+        $presta_callback_url = $api_url . '/callback.php?secret=' . $callback_secret;
+        $base_url = preg_replace('/https?:\/\//', '', $api_url);
+        $available_xpub = '';
+        $partial_match = '';
+        //Go through all xpubs on the server and examine their callback url
+        foreach ($response_body as $one_response) {
+            $server_callback_url = isset($one_response->callback) ? $one_response->callback : '';
+            $server_base_url = preg_replace('/https?:\/\//', '', $server_callback_url);
+            $xpub = isset($one_response->address) ? $one_response->address : '';
+            if (!$server_callback_url) {
+                // No callback
+                $available_xpub = $xpub;
+            } elseif ($server_callback_url == $presta_callback_url) {
+                // Exact match
+                return '';
+            } elseif (strpos($server_base_url, $base_url) === 0) {
+                // Partial Match - Only secret or protocol differ
+                $partial_match = $xpub;
+            }
+        }
+        // Use the available xpub
+        if ($partial_match || $available_xpub) {
+            $update_xpub = $partial_match ? $partial_match : $available_xpub;
+            $this->updateCallback($presta_callback_url, $crypto, $update_xpub);
+            return '';
+        }
+        // No match and no empty callback
+        $error_str = $this->l("Please add a new store on blockonomics' website");
         return $error_str;
     }
 
+    public function updateCallback($callback_url, $crypto, $xpub)
+    {
+        $set_callback_url = $this->getServerAPIURL($crypto, '/api/update_callback');
+        $post_content =
+        '{"callback": "' .
+        $callback_url .
+        '", "xpub": "' .
+        $xpub .
+        '"}';
+        $this->doCurlCall($set_callback_url, $post_content);
+    }
+
+    public function getCallbacks($crypto)
+    {
+        $path = '/api/address?&no_balance=true&only_xpub=true&get_callback=true';
```

Comment:
Declaring path like this is ok. But we are uncessary creating a new variable which is a constant on each function call. This is why programming guides recommend using way to declare a *global/class scope const string* variable. This way memory for this variable is allocated only once

Also it helps to have all these const strings at one place. Similar to the translate string that we are declaring

Reply:
True, I've added defined the strings as _global/class scope const string_. I also removed three lines of `Configuration::deleteByName` for variables that never were introduced.

As a side note, the WooCommerce plugin can be cleaned from code duplication in a similar way.